### PR TITLE
schedule: reorganize schedule package

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Part of modularizing PD.

### What is changed and how it works?
This PR contains some changes:
1. use an individual package for `operator`, `filter` and `selector`.
2. remove replica.go
3. rename `OperatorStep`, `OperatorHistory`, `OperatorKind`, `FilterTarget` and `FilterSource` to `OpStep`, `OpHistory`, `OpKind`, `Target` and `Source` respectively.
4. abstract some interfaces from `Cluster`, such as `StoreSetInformer`,  `StoreSetController` and `RegionStatInformer` and put them into the proper package.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: pkg/testutil/operator_check.go [hunk 1]
```diff
@@ -15,71 +15,71 @@ package testutil
 
 import (
 	check "github.com/pingcap/check"
-	"github.com/pingcap/pd/server/schedule"
+	"github.com/pingcap/pd/server/schedule/operator"
 )
 
 // CheckAddPeer checks if the operator is to add peer on specified store.
-func CheckAddPeer(c *check.C, op *schedule.Operator, kind schedule.OperatorKind, storeID uint64) {
+func CheckAddPeer(c *check.C, op *operator.Operator, kind operator.OpKind, storeID uint64) {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply pkg/testutil/operator_check.go]: For now, we already have some methods like `Step` and `Kind` for `Operator`. How about changing these methods to `GetStep` and `GetKind`?

File: server/schedule/filter/filters.go [hunk 2]
```diff
@@ -27,30 +28,30 @@ import (
 type Filter interface {
 	Type() string
 	// Return true if the store should not be used as a source store.
-	FilterSource(opt Options, store *core.StoreInfo) bool
+	Source(opt opt.Options, store *core.StoreInfo) bool
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedule/filter/filters.go]: Yes, absolutely. I will do it in another PR.

File: server/core/basic_cluster.go [hunk 3]
```diff
@@ -123,3 +123,33 @@ func (bc *BasicCluster) DeleteStore(store *StoreInfo) {
 func (bc *BasicCluster) PutRegion(region *RegionInfo) {
 	bc.Regions.SetRegion(region)
 }
+
+// RegionSetInformer provides access to a shared informer of regions.
+type RegionSetInformer interface {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/core/basic_cluster.go]: What's your opinion? /cc @nolouch

File: server/checker/namespace_checker.go [hunk 4]
```diff
@@ -19,20 +19,23 @@ import (
...
+	"github.com/pingcap/pd/server/schedule/operator"
+	"github.com/pingcap/pd/server/schedule/selector"
 	"go.uber.org/zap"
 )
 
 // NamespaceChecker ensures region to go to the right place.
 type NamespaceChecker struct {
 	cluster    schedule.Cluster
-	filters    []schedule.Filter
+	filters    []filter.Filter
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/checker/namespace_checker.go]: Yes. But I don't have any better idea either.

File: server/core/basic_cluster.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nolouch) [reply server/core/basic_cluster.go]: I think we should classify the roles, Like `Informer`, `List`?

**Comment by disksing** [comment]
I think we can merge it then consider name issues later.

File: server/checker/namespace_checker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (shafreeck) [reply server/checker/namespace_checker.go]: Maybe it is not regarded as verbose. If it is, go lint will tell us.

**Pull request merged**